Sources/CellViewModel.swift

```diff
@@ -72,7 +72,11 @@ extension CellViewModel {
 
     /// Returns a type-erased version of this view model.
     public func eraseToAnyViewModel() -> AnyCellViewModel {
-        AnyCellViewModel(self)
+        if let erasedViewModel = self as? AnyCellViewModel {
+            return erasedViewModel
+        } else {
+            return AnyCellViewModel(self)
+        }
```

Comment:
Instead of this check, I think we can implement this function on `AnyCellViewModel`

```swift
public struct AnyCellViewModel: CellViewModel {
    // existing implementation ....

    /// :nodoc:
    public func eraseToAnyViewModel() -> Self {
        self
    }
}
```

Reply:
If implement this, `func eraseToAnyViewModel()` should add to `protocol CellViewModel`, shell we do it?

Reply:
We could, but we don't strictly *need* to add it to the protocol. 

Let's **not** add it for now. If we find a reason that we need to do this later, we can. 

Thanks for asking! 😊

Reply:
If `func eraseToAnyViewModel()` not in `protocol CellViewModel`:

```swift
let viewModel = FakeTextCellViewModel() // viewModel.reuseIdentifier == "FakeTextCellViewModel"
let erased = viewModel.eraseToAnyViewModel() // erased.reuseIdentifier == "FakeTextCellViewModel"
let erased3 = viewModel.eraseToAnyViewModel().eraseToAnyViewModel() // erased3.reuseIdentifier == "FakeTextCellViewModel"
let erased4 = (viewModel.eraseToAnyViewModel() as (any CellViewModel)).eraseToAnyViewModel() // erased4.reuseIdentifier == "AnyCellViewModel"
```

Reply:
The same problem also occurs in `init<T>(_ viewModel: T) where T : CellViewModel`, `viewModel` may already be `AnyCellViewModel`, how should we deal with it?

Reply:
I think we can keep the implementation of `func eraseToAnyViewModel()` and change `init<T>(_ viewModel: T) where T : CellViewModel` to `fileprivate`.

Reply:
> If `func eraseToAnyViewModel()` not in `protocol CellViewModel`, it uses implementation in `CellViewModel` not in `AnyCellViewModel` when type is `any CellViewModel`:
> 
> ```swift
> let viewModel = FakeTextCellViewModel() // viewModel.reuseIdentifier == "FakeTextCellViewModel"
> let erased = viewModel.eraseToAnyViewModel() // erased.reuseIdentifier == "FakeTextCellViewModel"
> let erased3 = viewModel.eraseToAnyViewModel().eraseToAnyViewModel() // erased3.reuseIdentifier == "FakeTextCellViewModel"
> let erased4 = (viewModel.eraseToAnyViewModel() as (any CellViewModel)).eraseToAnyViewModel() // erased4.reuseIdentifier == "AnyCellViewModel"
> ```

Ohhh. Good catch. Ok, yes let's add `eraseToAnyViewModel()` to the `CellViewModel` protocol and let's add what you have here to the unit tests. 😊 

----

> The same problem also occurs in `init<T>(_ viewModel: T) where T : CellViewModel`, `viewModel` may already be `AnyCellViewModel`, how should we deal with it?

Doesn't the change above fix this? If we update `CellViewModel` protocol, then this is fixed, right?

Reply:
Unfortunately, even if we add `eraseToAnyViewModel()` to `CellViewModel` and add different implementations for `CellViewModel` and `AnyCellViewModel`, we still need to determine whether `viewModel` is `AnyCellViewModel` in `init()`.

---

```swift
public protocol CellViewModel {
    func eraseToAnyViewModel() -> AnyCellViewModel
}

extension CellViewModel {
    public func eraseToAnyViewModel() -> AnyCellViewModel {
        AnyCellViewModel(self)
    }
}

public struct AnyCellViewModel: CellViewModel {
    public func eraseToAnyViewModel() -> AnyCellViewModel {
        self
    }

    public init<T: CellViewModel>(_ viewModel: T) {
        if let erasedViewModel = viewModel as? AnyCellViewModel {
            self = erasedViewModel
            return
        }
        // original implementation
    }
}
```

Reply:
I think we could implement this:

```swift
extension CellViewModel {
    public func eraseToAnyViewModel() -> AnyCellViewModel {
        if let erasedViewModel = self as? AnyCellViewModel {
            return erasedViewModel
        } else {
            return AnyCellViewModel(self)
        }
    }
}

public struct AnyCellViewModel: CellViewModel {
    public init<T: CellViewModel>(_ viewModel: T) {
        if let erasedViewModel = viewModel as? AnyCellViewModel {
            self = erasedViewModel
            return
        }
        // original implementation
    }
}

Reply:
@nuomi1 Oh good catch. 💯  Yes, let's do what you've suggested here:

> I think we could implement this:
> 
> ```swift
> extension CellViewModel {
>     public func eraseToAnyViewModel() -> AnyCellViewModel {
>         if let erasedViewModel = self as? AnyCellViewModel {
>             return erasedViewModel
>         } else {
>             return AnyCellViewModel(self)
>         }
>     }
> }
> 
> public struct AnyCellViewModel: CellViewModel {
>     public init<T: CellViewModel>(_ viewModel: T) {
>         if let erasedViewModel = viewModel as? AnyCellViewModel {
>             self = erasedViewModel
>             return
>         }
>         // original implementation
>     }
> }
> ```

Reply:
Done.